File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/RtpSenderProxy.kt
package com.cloudwebrtc.webrtc.proxy

import com.cloudwebrtc.webrtc.exception.ReplaceSenderTrackException
import org.webrtc.RtpReceiver
import org.webrtc.RtpSender

/**
 * Wrapper around an [RtpSender].
 *
 * @param sender  Actual underlying [RtpSender].
 */
class RtpSenderProxy(sender: RtpSender) : Proxy<RtpSender> {
    /**
     * Actual underlying [RtpReceiver].
     */
    override var obj: RtpSender = sender

    /**
     * [MediaStreamTrackProxy] of this [RtpReceiverProxy].
     */
    private var track: MediaStreamTrackProxy? = null

    init {
        syncWithObject()
    }

    override fun syncWithObject() {
        syncMediaStreamTrack()
    }

    /**
     * Replaces [MediaStreamTrackProxy] of the underlying [RtpSender] with the
     * provided one.
     *
     * @param t  [MediaStreamTrackProxy] which will be set to the underlying
     *           [RtpSender].
     */
    fun replaceTrack(t: MediaStreamTrackProxy?) {
        track = t
        val isSuccessful = obj.setTrack(t?.obj, false)
        if (!isSuccessful) {
            throw ReplaceSenderTrackException()
        }
    }

    /**
     * Synchronizes the [MediaStreamTrackProxy] of this [RtpSenderProxy] with
     * the underlying [RtpSender].
     */
    private fun syncMediaStreamTrack() {
        val newSenderTrack = obj.track()
        if (newSenderTrack == null) {
            track = null
        } else {
            if (track == null) {
                track = MediaStreamTrackProxy(newSenderTrack)
            } else {
                track!!.replace(newSenderTrack)
            }
        }
    }
}

File: ios/Classes/proxy/RtpSenderProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpSender`, powering it with additional API.
class RtpSenderProxy {
  /// Actual underlying `RTCRtpSender`.
  private var sender: RTCRtpSender

  /// `MediaStreamTrackProxy` of this `RtpSenderProxy`.
  private var track: MediaStreamTrackProxy?

  /// Initializes a new `RtpSenderProxy` for the provided `RTCRtpSender`.
  init(sender: RTCRtpSender) {
    self.sender = sender
    self.syncMediaStreamTrack()
  }

  /// Returns ID of this `RtpSenderProxy`.
  func id() -> String {
    self.sender.senderId
  }

  /// Replaces the `MediaStreamTrackProxy` of the underlying `RTCRtpSender` with
  /// the provided one.
  func replaceTrack(t: MediaStreamTrackProxy?) {
    self.track = t
    self.sender.track = t?.obj()
  }

  /// Returns `RTCRtpParameters` of the underlying `RTCRtpSender`.
  func getParameters() -> RTCRtpParameters {
    return self.sender.parameters
  }

  /// Sets `RTCRtpParameters` of the underlying `RTCRtpSender` with the provided
  /// one.
  func setParameters(params: RTCRtpParameters) {
    self.sender.parameters = params
  }

  /// Synchronizes the `MediaStreamTrackProxy` of this `RtpSenderProxy` with the
  /// underlying `RTCRtpSender`.
  func syncMediaStreamTrack() {
    let newTrack = self.sender.track
    if newTrack == nil {
      self.track = nil
    } else {
      if self.track == nil {
        self.track = MediaStreamTrackProxy(
          track: newTrack!,
          deviceId: nil,
          source: nil
        )
      }
    }
  }
}
